---
title: "Shifts in mutation effects among variant backgrounds"
author: "Tyler Starr"
date: "10/11/2021"
output:
  github_document:
    toc: true
    html_preview: false
editor_options: 
  chunk_output_type: inline
---
This notebook analyzes sites whose mutation effects deviate most strongly among the variant RBD backgrounds.

```{r setup, message=FALSE, warning=FALSE, error=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(dev.args = list(png = list(type = "cairo")))

#list of packages to install/load
packages = c("yaml","data.table","tidyverse","gridExtra","egg","bio3d")
#install any packages not already installed
installed_packages <- packages %in% rownames(installed.packages())
if(any(installed_packages == F)){
  install.packages(packages[!installed_packages])
}
#load packages
invisible(lapply(packages, library, character.only=T))

#read in config file
config <- read_yaml("config.yaml")

#make output directory
if(!file.exists(config$epistatic_shifts_dir)){
  dir.create(file.path(config$epistatic_shifts_dir))
}

#make pdb output directory
if(!file.exists(paste(config$epistatic_shifts_dir,"/pdbs/",sep=""))){
  dir.create(file.path(paste(config$epistatic_shifts_dir,"/pdbs/",sep="")))
}
```
Session info for reproducing environment:
```{r print_sessionInfo}
sessionInfo()
```

Define colorblind-friendly palette
```{r define_color_palette}
# The palette with grey:
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", 
               "#0072B2", "#D55E00", "#CC79A7")
# The palette with black
cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", 
                "#0072B2", "#D55E00", "#CC79A7")
```

## Setup

Read in tables of mutant measurements.

```{r input_data}
dt <- data.table(read.csv(file=config$final_variant_scores_mut_file,stringsAsFactors=F))

```

## Calculate site-wise mean absolute error, slope from correlation in mutation effects

For each pair of backgrounds, at each site I want to compute the mean absolute error in affinities across the 20 amino acids measured (wildtype and 19 mutants). Deal with measurements in the 5-7 range a few different ways, as noted in the # comments in code


```{r setup_table}
#define a censored bind measurement that collapses anything <6 to 6 (a lot more noise in the 5-6 range due to concentration range)
dt[,bind_cens := bind]
dt[!is.na(bind) & bind<6,bind_cens:=6]

#define a minbc bind measurement that requires three barcodes be averaged for a final determination, otherwise change to NA
dt[,bind_min3bc := bind]
dt[n_bc_bind < 3, bind_min3bc := NA]

dt[,bind_min5bc := bind]
dt[n_bc_bind < 5, bind_min5bc := NA]

#define a function for computing J-S divergence between two affinity vectors
JS <- function(vec1,vec2){
  vec1_pair <- vec1[!is.na(vec1) & !is.na(vec2)]
  vec2_pair <- vec2[!is.na(vec1) & !is.na(vec2)]
  pi1 <- exp(vec1_pair)/sum(exp(vec1_pair))
  pi2 <- exp(vec2_pair)/sum(exp(vec2_pair))
  n <- 0.5 * (pi1+pi2)
  JS <- 0.5 * (sum(pi1*log(pi1/n)) + sum(pi2*log(pi2/n)))
  return(sqrt(JS))
}

#first, for bind measurements
#data table for storing difference in correlation in profiles between bg pairs at each site
#generate table with all combinations of bg_1 and bg_2 for each site
diffs_bind <- data.table(expand.grid(site=unique(dt$position),bg_2=c("Wuhan_Hu_1","E484K","N501Y","B1351","Delta"),bg_1=c("Wuhan_Hu_1","E484K","N501Y","B1351","Delta")))

#remove duplicates -- either bg_1 and _2 the same, or combinations where the _2 _1 combo is already present in the _1 _2 orientation
diffs_bind <- diffs_bind[bg_1 != bg_2,]
diffs_bind <- diffs_bind[bg_1 == "Wuhan_Hu_1" | (bg_1=="E484K" & bg_2 %in% c("N501Y","B1351","Delta")) | (bg_1=="N501Y" & bg_2 %in% c("B1351","Delta")) | bg_1=="B1351" & bg_2=="Delta"]

#loop through and compute mean absolute error, correlation coefficients, and regression slope
diffs_bind$cor <- as.numeric(NA)
diffs_bind$slope <- as.numeric(NA)
diffs_bind$rho <- as.numeric(NA)
diffs_bind$MAE <- as.numeric(NA) #score of mean average error from bind_cens, including values at boundary condition = 6
diffs_bind$MAE_cens <- as.numeric(NA) #score of mean average error from bind_cens, excluding amino acids that are at the boudnary condition = 6 *in both bgs*
diffs_bind$JSD <- as.numeric(NA) #jensen-shannon divergence, from raw bind values (lower limit 5)
diffs_bind$JSD_min3bc <- as.numeric(NA) #jensen-shannon divergence, require a minimum of 3 bcs averaged
diffs_bind$JSD_min5bc <- as.numeric(NA) #jensen-shannon divergence, require a minimum of 5 bcs averaged
diffs_bind$JSD_cens <- as.numeric(NA) #jensen-shannon divergence, from bind_cens values (lower limit 6)
for(i in 1:nrow(diffs_bind)){
  x_uncens <- dt[target==diffs_bind[i,bg_1] & position==diffs_bind[i,site],bind]
  y_uncens <- dt[target==diffs_bind[i,bg_2] & position==diffs_bind[i,site],bind]
  x <- dt[target==diffs_bind[i,bg_1] & position==diffs_bind[i,site],bind_cens]
  y <- dt[target==diffs_bind[i,bg_2] & position==diffs_bind[i,site],bind_cens]
  x_cens <- x[(x>6 | y>6)]
  y_cens <- y[(x>6 | y>6)]
  x_min3bc <- dt[target==diffs_bind[i,bg_1] & position==diffs_bind[i,site],bind_min3bc]
  y_min3bc <- dt[target==diffs_bind[i,bg_2] & position==diffs_bind[i,site],bind_min3bc]
  x_min5bc <- dt[target==diffs_bind[i,bg_1] & position==diffs_bind[i,site],bind_min5bc]
  y_min5bc <- dt[target==diffs_bind[i,bg_2] & position==diffs_bind[i,site],bind_min5bc]
  diffs_bind[i,cor:=cor(x,y,use="complete.obs",method="pearson")]
  diffs_bind[i,slope:=lm(y~x)$coefficients["x"]]
  diffs_bind[i,rho:=cor(x,y,use="complete.obs",method="spearman")]
  diffs_bind[i,MAE:=mean(abs(lm(y~x)$residuals),na.rm=T)]
  diffs_bind[i,MAE_cens:=mean(abs(lm(y_cens~x_cens)$residuals),na.rm=T)]
  diffs_bind[i,JSD := JS(x_uncens,y_uncens)]
  diffs_bind[i,JSD_min3bc := JS(x_min3bc,y_min3bc)]
  diffs_bind[i,JSD_min5bc := JS(x_min3bc,y_min5bc)]
  diffs_bind[i,JSD_cens := JS(x,y)]
}

#repeat for expr measurements
#data table for storign difference in correlation in profiles between bg pairs at each site
#generate table with all combinations of bg_1 and bg_2 for each site
diffs_expr <- data.table(expand.grid(site=unique(dt$position),bg_2=c("Wuhan_Hu_1","E484K","N501Y","B1351","Delta"),bg_1=c("Wuhan_Hu_1","E484K","N501Y","B1351","Delta")))

#remove duplicates -- either bg_1 and _2 the same, or combinations where the _2 _1 combo is already present in the _1 _2 orientation
diffs_expr <- diffs_expr[bg_1 != bg_2,]
diffs_expr <- diffs_expr[bg_1 == "Wuhan_Hu_1" | (bg_1=="E484K" & bg_2 %in% c("N501Y","B1351","Delta")) | (bg_1=="N501Y" & bg_2 %in% c("B1351","Delta")) | bg_1=="B1351" & bg_2=="Delta"]

#loop through and compute mean absolute error and correlation coefficient r
diffs_expr$cor <- as.numeric(NA)
diffs_expr$slope <- as.numeric(NA)
diffs_expr$rho <- as.numeric(NA)
diffs_expr$MAE <- as.numeric(NA)
for(i in 1:nrow(diffs_expr)){
  x <- dt[target==diffs_expr[i,bg_1] & position==diffs_expr[i,site],expr]
  y <- dt[target==diffs_expr[i,bg_2] & position==diffs_expr[i,site],expr]
  diffs_expr[i,cor:=cor(x,y,use="complete.obs",method="pearson")]
  diffs_expr[i,slope:=lm(y~x)$coefficients["x"]]
  diffs_expr[i,rho:=cor(x,y,use="complete.obs",method="spearman")]
  diffs_expr[i,MAE:=mean(abs(lm(y~x)$residuals),na.rm=T)]
}

```

Plotting/visualizing:

Utility function: plot scatterplot showing affinity of each of the 20 amino acids in a pair of sites

```{r scatterplot_function}
plot_scatter <- function(site, bg1, bg2, MAE_slope=F, MAE=F, MAE_cens=F, JSD=T, JSD_min3bc=F, JSD_min5bc=F, rho=F, cor=F,cutoff=5,phenotype="bind"){
  x <- dt[target==bg1 & position==site,get(phenotype)]
  x_ref <- dt[target==bg1 & position==site & as.character(mutant)==as.character(wildtype),get(phenotype)]
  y <- dt[target==bg2 & position==site,get(phenotype)]
  y_ref <- dt[target==bg2 & position==site & as.character(mutant)==as.character(wildtype),get(phenotype)]
  x_cens <- x[(x>6 | y>6)]
  y_cens <- y[(x>6 | y>6)]
  x_min3bc <- dt[target==bg1 & position==site,bind_min3bc]
  y_min3bc <- dt[target==bg2 & position==site,bind_min3bc]
  x_min5bc <- dt[target==bg1 & position==site,bind_min5bc]
  y_min5bc <- dt[target==bg2 & position==site,bind_min5bc]
  chars <- dt[target==bg1 & position==site,mutant]
  cols <- rep("black",20); cols[which(x <= cutoff & y <= cutoff)] <- "orange"
  plot(x,y, xlim=c(4.5,12),ylim=c(4.5,12),pch=chars,xlab=paste(bg1,phenotype),ylab=paste(bg2,phenotype),col=cols,main=paste("site",site))
  abline(v=x_ref,lty=2,col="red")
  abline(h=y_ref,lty=2,col="red")
  if(MAE_slope==T){
    val1 <- mean(abs(lm(y~x)$residuals),na.rm=T)
    val2 <- lm(y~x)$coefficients["x"]
    legend("topleft",bty="n",cex=1,legend=paste("MAE:",format(val1,digits=3),"\nslope:",format(val2,digits=2)))
  }else if(MAE==T){
    val <- mean(abs(lm(y~x)$residuals),na.rm=T)
    legend("topleft",bty="n",cex=1,legend=paste("MAE:",format(val,digits=3)))
  }else if(MAE_cens==T){
    val <- mean(abs(lm(y_cens~x_cens)$residuals),na.rm=T)
    legend("topleft",bty="n",cex=1,legend=paste("MAE (cens):",format(val,digits=3)))
  }else if(JSD==T){
    val <- JS(x,y)
    legend("topleft",bty="n",cex=1,legend=paste("divergence:",format(val,digits=3)))
  }else if(JSD_min3bc==T){
    val <- JS(x_min3bc,y_min3bc)
    legend("topleft",bty="n",cex=1,legend=paste("divergence (min 3bc):",format(val,digits=3)))
  }else if(JSD_min5bc==T){
    val <- JS(x_min5bc,y_min5bc)
    legend("topleft",bty="n",cex=1,legend=paste("divergence (min 5bc):",format(val,digits=3)))
  }else if(rho==T){
    val <- cor(x,y,use="complete.obs",method="spearman")
    legend("topleft",bty="n",cex=1,legend=paste("rho:",format(val,digits=3)))
  }else if(cor==T){
    val <- cor(x,y,use="complete.obs",method="pearson")
    legend("topleft",bty="n",cex=1,legend=paste("r:",format(val,digits=3)))
  }
}

```

```{r example_correlation, echo=T, fig.width=8, fig.height=8, fig.align="center", dpi=300,dev="png"}
par(mfrow=c(2,2))
plot_scatter(site=484,"Wuhan_Hu_1","N501Y")
plot_scatter(site=501,"Wuhan_Hu_1","N501Y")
plot_scatter(site=403,"Wuhan_Hu_1","N501Y")
plot_scatter(site=498,"Wuhan_Hu_1","N501Y")

```
```{r corrs_B1351_WH1_446-loop, echo=T, fig.width=8, fig.height=8, fig.align="center", dpi=300,dev="png"}
par(mfrow=c(2,2))
plot_scatter(site=446,"Wuhan_Hu_1","B1351")
plot_scatter(site=447,"Wuhan_Hu_1","B1351")
plot_scatter(site=448,"Wuhan_Hu_1","B1351")
plot_scatter(site=449,"Wuhan_Hu_1","B1351")

```

```{r corrs_452, echo=T, fig.width=8, fig.height=8, fig.align="center", dpi=300,dev="png"}
par(mfrow=c(2,2))
plot_scatter(site=452,"Wuhan_Hu_1","B1351")
plot_scatter(site=452,"Wuhan_Hu_1","N501Y")
plot_scatter(site=452,"Wuhan_Hu_1","E484K")
plot_scatter(site=452,"E484K","B1351")

```

TODO: Utility funciton: plot heatmap showing mutant effects profiles of the five backgrounds on a site-by-site basis

TODO: line plots of JSD of focal bg versus other across sites
```{r line_plots_JSD_v_WH1, echo=T, fig.width=12, fig.height=4, fig.align="center", dpi=300,dev="png"}
#define focal bg for others to compare to
bg <- "Wuhan_Hu_1"
temp <- diffs_bind[bg_1==bg | bg_2 == bg]
temp$target <- as.character(NA)
for(i in 1:nrow(temp)){
  if(temp[i,bg_1]!=bg){
    temp[i,target:=temp[i,bg_1]]
  }else if(temp[i,bg_2]!=bg){
    temp[i,target:=temp[i,bg_2]]
  }
}

#define colors for each bg
group.colors <- c("Wuhan_Hu_1" = cbPalette[1], "N501Y" = cbPalette[3], "E484K" = cbPalette[5], "B1351"=cbPalette[6], "Delta"=cbPalette[7])

#define some epitope classes for adding highlights
label_df <- data.frame(xmin=c(382, 482, 454, 442, 416),
                       xmax=c(387, 487, 457, 451, 418),
                       antibody_class=c("class 4", "class 2", "class 1", "class 3", "class 1"),
                       label=c("383-386","483-486","455-456","443-450","417"))

ggplot(data=temp, aes(x=site, y=JSD, color=target))+
  geom_rect(data=label_df, aes(x=NULL, y=NULL, color=NULL,xmin=xmin, xmax=xmax, ymin=0,ymax=1.1*max(temp$JSD,na.rm=T),fill=antibody_class), alpha=0.2)+
  geom_line(size=0.5)+
  scale_color_manual(values=group.colors)+
  theme_classic()+
  scale_x_continuous(expand=c(0.01,0.01),breaks=c(331,seq(335,530,by=5)))+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold",size=10))+
  ylab("JS divergence versus Wuhan-Hu-1")

invisible(dev.print(pdf, paste(config$epistatic_shifts_dir,"/JSD_v_WH1.pdf",sep=""),useDingbats=F))
```

Same but require minimum 3 bc for a measurement
```{r line_plots_JSD_v_WH1_min3bc, echo=T, fig.width=12, fig.height=4, fig.align="center", dpi=300,dev="png"}
ggplot(data=temp, aes(x=site, y=JSD_min3bc, color=target))+
  geom_rect(data=label_df, aes(x=NULL, y=NULL, color=NULL,xmin=xmin, xmax=xmax, ymin=0,ymax=1.1*max(temp$JSD_min3bc,na.rm=T),fill=antibody_class), alpha=0.2)+
  geom_line(size=0.5)+
  scale_color_manual(values=group.colors)+
  theme_classic()+
  scale_x_continuous(expand=c(0.01,0.01),breaks=c(331,seq(335,530,by=5)))+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold",size=10))+
  ylab("JS divergence versus Wuhan-Hu-1")

invisible(dev.print(pdf, paste(config$epistatic_shifts_dir,"/JSD_v_WH1_min3bc.pdf",sep=""),useDingbats=F))
```

Separate facet per bg comparison

```{r line_plots_JSD_v_WH1_faceted, echo=T, fig.width=12, fig.height=12, fig.align="center", dpi=300,dev="png"}
ggplot(data=temp, aes(x=site, y=JSD, color=target))+
  geom_rect(data=label_df, aes(x=NULL, y=NULL, color=NULL,xmin=xmin, xmax=xmax, ymin=0,ymax=1.1*max(temp$JSD,na.rm=T),fill=antibody_class), alpha=0.2)+
  geom_line(size=0.75)+
  scale_color_manual(values=group.colors)+
  theme_classic()+
  scale_x_continuous(expand=c(0.01,0.01),breaks=c(331,seq(335,530,by=5)))+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold",size=10))+
  ylab("JS divergence versus Wuhan-Hu-1")+
  facet_wrap(~target,ncol=1)

invisible(dev.print(pdf, paste(config$epistatic_shifts_dir,"/JSD_v_WH1_faceted.pdf",sep=""),useDingbats=F))
```

Again requiring min 3 bc

```{r line_plots_JSD_v_WH1_faceted_min3bc, echo=T, fig.width=12, fig.height=12, fig.align="center", dpi=300,dev="png"}
ggplot(data=temp, aes(x=site, y=JSD_min3bc, color=target))+
  geom_rect(data=label_df, aes(x=NULL, y=NULL, color=NULL,xmin=xmin, xmax=xmax, ymin=0,ymax=1.1*max(temp$JSD_min3bc,na.rm=T),fill=antibody_class), alpha=0.2)+
  geom_line(size=0.75)+
  scale_color_manual(values=group.colors)+
  theme_classic()+
  scale_x_continuous(expand=c(0.01,0.01),breaks=c(331,seq(335,530,by=5)))+
  theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.6,face="bold",size=10))+
  ylab("JS divergence versus Wuhan-Hu-1")+
  facet_wrap(~target,ncol=1)

invisible(dev.print(pdf, paste(config$epistatic_shifts_dir,"/JSD_v_WH1_faceted_min3bc.pdf",sep=""),useDingbats=F))
```

## Map divergence to pdb structure
```{r map_JSD_to_pdb}
pdb_wh1 <- read.pdb(file=config$pdb_6m0j)

#iterate through backgrounds, output a pdb comparing its divergence to WH1 (using min3bc)
for(s in c("E484K","N501Y","B1351","Delta")){
  b <- rep(0, length(pdb_wh1$atom$b))
  for(i in 1:nrow(pdb_wh1$atom)){
    if(pdb_wh1$atom$chain[i]=="E"){
      res <- pdb_wh1$atom$resno[i]
      JSD <- diffs_bind[bg_1=="Wuhan_Hu_1" & bg_2==s & site==res, JSD_min3bc]
      if(length(JSD)>0){
        b[i] <- JSD
      }
    }
  }
  write.pdb(pdb=pdb_wh1, file=paste(config$epistatic_shifts_dir,"/pdbs/",s,"_v_WH1_JSD-min3bc.pdb",sep=""), b=b)
}

```


## Nonspecific compensation

Additivity of 417/484/501 muts? Build out cycle of mutants from the different places where each mutant is measured (e.g. WH1 is also measured in E484K as the K484E mut)

```{r additivity_triple_mut_cycle_bind, echo=T, fig.width=5, fig.height=4, fig.align="center", dpi=300,dev="png"}
cycle_bind <- data.table()

cycle_bind$geno <- "Wuhan_Hu_1"
cycle_bind$bind <- dt[target=="Wuhan_Hu_1" & position==331 & wildtype==mutant,bind]
cycle_bind$n_bc <- dt[target=="Wuhan_Hu_1" & position==331 & wildtype==mutant,n_bc_bind]
cycle_bind$nmut <- 0


cycle_bind <- rbind(cycle_bind,list("Wuhan_Hu_1",dt[target=="E484K" & position==484 & mutant=="E",bind],dt[target=="E484K" & position==484 & mutant=="E",n_bc_bind],0))
cycle_bind <- rbind(cycle_bind,list("Wuhan_Hu_1",dt[target=="N501Y" & position==501 & mutant=="N",bind],dt[target=="N501Y" & position==501 & mutant=="N",n_bc_bind],0))


#K417N
cycle_bind <- rbind(cycle_bind,list("K417N",dt[target=="Wuhan_Hu_1" & position==417 & mutant=="N",bind],dt[target=="Wuhan_Hu_1" & position==417 & mutant=="N",n_bc_bind],1))

#E484K
cycle_bind <- rbind(cycle_bind,list("E484K",dt[target=="Wuhan_Hu_1" & position==484 & mutant=="K",bind],dt[target=="Wuhan_Hu_1" & position==484 & mutant=="K",n_bc_bind],1))
cycle_bind <- rbind(cycle_bind,list("E484K",dt[target=="E484K" & position==484 & mutant==wildtype,bind],dt[target=="E484K" & position==484 & mutant==wildtype,n_bc_bind],1))

#N501Y
cycle_bind <- rbind(cycle_bind,list("N501Y",dt[target=="Wuhan_Hu_1" & position==501 & mutant=="Y",bind],dt[target=="Wuhan_Hu_1" & position==501 & mutant=="Y",n_bc_bind],1))
cycle_bind <- rbind(cycle_bind,list("N501Y",dt[target=="N501Y" & position==484 & mutant==wildtype,bind],dt[target=="N501Y" & position==484 & mutant==wildtype,n_bc_bind],1))

#K417N/E484K
cycle_bind <- rbind(cycle_bind,list("K417N/E484K",dt[target=="E484K" & position==417 & mutant=="N",bind],dt[target=="E484K" & position==417 & mutant=="N",n_bc_bind],2))
cycle_bind <- rbind(cycle_bind,list("K417N/E484K",dt[target=="B1351" & position==501 & mutant=="N",bind],dt[target=="B1351" & position==501 & mutant=="N",n_bc_bind],2))

#K417N/N501Y
cycle_bind <- rbind(cycle_bind,list("K417N/N501Y",dt[target=="N501Y" & position==417 & mutant=="N",bind],dt[target=="N501Y" & position==417 & mutant=="N",n_bc_bind],2))
cycle_bind <- rbind(cycle_bind,list("K417N/N501Y",dt[target=="B1351" & position==484 & mutant=="E",bind],dt[target=="B1351" & position==484 & mutant=="E",n_bc_bind],2))

#E484K/N501Y
cycle_bind <- rbind(cycle_bind,list("E484K/N501Y",dt[target=="E484K" & position==501 & mutant=="Y",bind],dt[target=="E484K" & position==501 & mutant=="Y",n_bc_bind],2))
cycle_bind <- rbind(cycle_bind,list("E484K/N501Y",dt[target=="N501Y" & position==484 & mutant=="K",bind],dt[target=="N501Y" & position==484 & mutant=="K",n_bc_bind],2))
cycle_bind <- rbind(cycle_bind,list("E484K/N501Y",dt[target=="B1351" & position==417 & mutant=="K",bind],dt[target=="B1351" & position==417 & mutant=="K",n_bc_bind],2))

#B1351
cycle_bind <- rbind(cycle_bind,list("B1351",dt[target=="B1351" & position==417 & mutant==wildtype,bind],dt[target=="B1351" & position==417 & mutant==wildtype,n_bc_bind],3))

plot(cycle_bind$nmut,cycle_bind$bind,pch=19,xlab="number mutations",ylab="binding affinity (log10(Kd))")

invisible(dev.print(pdf, paste(config$epistatic_shifts_dir,"/B1351_epistasis_cycle_bind.pdf",sep=""),useDingbats=F))
```

```{r additivity_triple_mut_cycle_expr, echo=T, fig.width=5, fig.height=4, fig.align="center", dpi=300,dev="png"}
cycle_expr <- data.table()

cycle_expr$geno <- "Wuhan_Hu_1"
cycle_expr$expr <- dt[target=="Wuhan_Hu_1" & position==331 & wildtype==mutant,expr]
cycle_expr$n_bc <- dt[target=="Wuhan_Hu_1" & position==331 & wildtype==mutant,n_bc_expr]
cycle_expr$nmut <- 0


cycle_expr <- rbind(cycle_expr,list("Wuhan_Hu_1",dt[target=="E484K" & position==484 & mutant=="E",expr],dt[target=="E484K" & position==484 & mutant=="E",n_bc_expr],0))
cycle_expr <- rbind(cycle_expr,list("Wuhan_Hu_1",dt[target=="N501Y" & position==501 & mutant=="N",expr],dt[target=="N501Y" & position==501 & mutant=="N",n_bc_expr],0))


#K417N
cycle_expr <- rbind(cycle_expr,list("K417N",dt[target=="Wuhan_Hu_1" & position==417 & mutant=="N",expr],dt[target=="Wuhan_Hu_1" & position==417 & mutant=="N",n_bc_expr],1))

#E484K
cycle_expr <- rbind(cycle_expr,list("E484K",dt[target=="Wuhan_Hu_1" & position==484 & mutant=="K",expr],dt[target=="Wuhan_Hu_1" & position==484 & mutant=="K",n_bc_expr],1))
cycle_expr <- rbind(cycle_expr,list("E484K",dt[target=="E484K" & position==484 & mutant==wildtype,expr],dt[target=="E484K" & position==484 & mutant==wildtype,n_bc_expr],1))

#N501Y
cycle_expr <- rbind(cycle_expr,list("N501Y",dt[target=="Wuhan_Hu_1" & position==501 & mutant=="Y",expr],dt[target=="Wuhan_Hu_1" & position==501 & mutant=="Y",n_bc_expr],1))
cycle_expr <- rbind(cycle_expr,list("N501Y",dt[target=="N501Y" & position==484 & mutant==wildtype,expr],dt[target=="N501Y" & position==484 & mutant==wildtype,n_bc_expr],1))

#K417N/E484K
cycle_expr <- rbind(cycle_expr,list("K417N/E484K",dt[target=="E484K" & position==417 & mutant=="N",expr],dt[target=="E484K" & position==417 & mutant=="N",n_bc_expr],2))
cycle_expr <- rbind(cycle_expr,list("K417N/E484K",dt[target=="B1351" & position==501 & mutant=="N",expr],dt[target=="B1351" & position==501 & mutant=="N",n_bc_expr],2))

#K417N/N501Y
cycle_expr <- rbind(cycle_expr,list("K417N/N501Y",dt[target=="N501Y" & position==417 & mutant=="N",expr],dt[target=="N501Y" & position==417 & mutant=="N",n_bc_expr],2))
cycle_expr <- rbind(cycle_expr,list("K417N/N501Y",dt[target=="B1351" & position==484 & mutant=="E",expr],dt[target=="B1351" & position==484 & mutant=="E",n_bc_expr],2))

#E484K/N501Y
cycle_expr <- rbind(cycle_expr,list("E484K/N501Y",dt[target=="E484K" & position==501 & mutant=="Y",expr],dt[target=="E484K" & position==501 & mutant=="Y",n_bc_expr],2))
cycle_expr <- rbind(cycle_expr,list("E484K/N501Y",dt[target=="N501Y" & position==484 & mutant=="K",expr],dt[target=="N501Y" & position==484 & mutant=="K",n_bc_expr],2))
cycle_expr <- rbind(cycle_expr,list("E484K/N501Y",dt[target=="B1351" & position==417 & mutant=="K",expr],dt[target=="B1351" & position==417 & mutant=="K",n_bc_expr],2))

#B1351
cycle_expr <- rbind(cycle_expr,list("B1351",dt[target=="B1351" & position==417 & mutant==wildtype,expr],dt[target=="B1351" & position==417 & mutant==wildtype,n_bc_expr],3))

plot(cycle_expr$nmut,cycle_expr$expr,pch=19,xlab="number mutations",ylab="expression (log(MFI))")

invisible(dev.print(pdf, paste(config$epistatic_shifts_dir,"/B1351_epistasis_cycle_expr.pdf",sep=""),useDingbats=F))
```

## Comparison of mutant accumulation on N501 versus Y501 backgrounds

Does affinity enhancement of N501Y enable more robustness to affinity-decreasing mutations on Y501 backgrounds?

Working with preliminary output generated by Will to enumerate phylogenetic substitution occurrence (so, not simply sequence counts) of mutations on N501 versus Y501 backgrounds. We can later integrate this as a script into the analysis pipeline

Read in counts, add affinity change information, generate vectors of mutant counts in each background, plot violin plots of distributions

```{r substitution_accrual_N501Y, echo=T, fig.width=5, fig.height=6, fig.align="center", dpi=300,dev="png"}
subs <- data.table(read.csv(file="data/background_count.csv"),stringsAsFactors = F)

#keep just RBD changes
subs <- subs[POS %in% seq(331,531),]

#remove rows with single counts? (also, check with Will that muts to tips are not being counted?)
subs <- subs[Count>1,]

#sort by bg and position for convenience
subs <- subs[order(subs$Background,subs$POS)]

subs$delta_bind <- as.numeric(NA)

#maybe only record subs from the "wildtype" state (WH1)?
for(i in 1:nrow(subs)){
  if(subs[i,WT] != "*" & subs[i,MUT] != "*"){
    if(subs[i,Background=="N501N"]){
      ref_bind <- dt[target=="Wuhan_Hu_1" & position==subs[i,POS] & mutant==as.character(subs[i,WT]),bind]
      mut_bind <- dt[target=="Wuhan_Hu_1" & position==subs[i,POS] & mutant==as.character(subs[i,MUT]),bind]
      subs[i,"delta_bind"] <- mut_bind - ref_bind
    }else if(subs[i,Background=="N501Y"]){
      ref_bind <- dt[target=="N501Y" & position==subs[i,POS] & mutant==as.character(subs[i,WT]),bind]
      mut_bind <- dt[target=="N501Y" & position==subs[i,POS] & mutant==as.character(subs[i,MUT]),bind]
      subs[i,"delta_bind"] <- mut_bind - ref_bind
    }
  }
}




#remove site 501 changes from these counts for now?
subs[POS == 501, delta_bind:=NA]

#build up vectors that include sub counts
delta_bind_N501 <- vector()
delta_bind_Y501 <- vector()

for(i in 1:nrow(subs)){
  if(subs[i,Background=="N501N"]){
    delta_bind_N501 <- c(delta_bind_N501,rep(subs[i,delta_bind],subs[i,Count]))
  }else if(subs[i,Background=="N501Y"]){
    delta_bind_Y501 <- c(delta_bind_Y501,rep(subs[i,delta_bind],subs[i,Count]))
  }
}

par(mfrow=c(2,1))
hist(delta_bind_N501, breaks=40, xlab="",main="N501",xlim=range(c(delta_bind_N501,delta_bind_Y501),na.rm=T))
hist(delta_bind_Y501, breaks=40, xlab="substitution effect on ACE2 affinity",main="Y501",xlim=range(c(delta_bind_N501,delta_bind_Y501),na.rm=T))

```




